velox/substrait/SubstraitToVeloxExpr.cpp

```diff
@@ -105,12 +108,269 @@ SubstraitVeloxExprConverter::toVeloxExpr(
+  if (typeCase == ::substrait::Expression_Literal::LiteralTypeCase::kList) {
+    if (listLiteral.list().values().size() == 0) {
+      return;
+    }
+    for (auto child : listLiteral.list().values()) {
+      listLiteralToFlatVector<kind>(child, vector, index);
+    }
+    return;
+  }
+
+  if constexpr (std::is_same_v<TCpp, StringView>) {
+    if (typeCase == ::substrait::Expression_Literal::LiteralTypeCase::kString) {
+      flatVector->set(index++, StringView(listLiteral.string()));
+    }
+    if (typeCase ==
+        ::substrait::Expression_Literal::LiteralTypeCase::kVarChar) {
+      flatVector->set(index++, StringView(listLiteral.var_char().value()));
+    }
+  } else if constexpr (
+      std::is_same_v<TCpp, Timestamp> || std::is_same_v<TCpp, Date> ||
+      std::is_same_v<TCpp, IntervalDayTime>) {
+  } else {
+    if (typeCase == ::substrait::Expression_Literal::LiteralTypeCase::kNull) {
+      vector->setNull(index++, true);
+    } else {
+      flatVector->set(index++, GetLiteralValue<TCpp>(listLiteral));
+    }
+  }
+  return;
+}
+
+VectorPtr SubstraitVeloxExprConverter::constructFlatVector(
+    const ::substrait::Expression::Literal& listLiteral,
+    const vector_size_t size,
+    const TypePtr& type) {
+  VELOX_CHECK(type->isPrimitiveType());
+  auto vector = BaseVector::create(type, size, pool_);
+
+  vector_size_t index = 0;
+  auto kind = vector->typeKind();
+  VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH(
+      listLiteralToFlatVector, kind, listLiteral, vector, index);
+  return vector;
+}
+
+ArrayVectorPtr SubstraitVeloxExprConverter::literalsToArrayVector(
+    const ::substrait::Expression::Literal& listLiteral) {
+  if (listLiteral.list().values().size() == 0) {
+    std::vector<variant> variants;
+    auto varArray = variant::array(variants);
+    ArrayVectorPtr arrayVector = variantArrayToVector(varArray.array(), pool_);
+    return arrayVector;
+  }
+  auto typeCase = listLiteral.list().values(0).literal_type_case();
+  switch (typeCase) {
+    case ::substrait::Expression_Literal::LiteralTypeCase::kBoolean:
+    case ::substrait::Expression_Literal::LiteralTypeCase::kI8:
+    case ::substrait::Expression_Literal::LiteralTypeCase::kI16:
+    case ::substrait::Expression_Literal::LiteralTypeCase::kI32:
+    case ::substrait::Expression_Literal::LiteralTypeCase::kFp32:
+    case ::substrait::Expression_Literal::LiteralTypeCase::kI64:
+    case ::substrait::Expression_Literal::LiteralTypeCase::kFp64:
+    case ::substrait::Expression_Literal::LiteralTypeCase::kString:
+    case ::substrait::Expression_Literal::LiteralTypeCase::kNull:
+    case ::substrait::Expression_Literal::LiteralTypeCase::kVarChar: {
+      std::vector<variant> variants;
+      variants.reserve(listLiteral.list().values().size());
+      std::optional<TypePtr> literalType = std::nullopt;
+      for (const auto& literal : listLiteral.list().values()) {
+        auto veloxVariant = toVeloxExpr(literal)->value();
```

Comment:
@zhejiangxiaomai 

I find it hard to follow the code. It looks like a lot of redundant conversions are happening. Here, I'm seeing that we are calling toVeloxExpr for each array element, which internal does a type-switch (redundant), constructs a variant, which then needs to be converted to a Vector. Can we streamline this processing?

Reply:
Yes, I will find a better way to streamline this processing.

Reply:
OK, I will try constructFlatVector  function.

Reply:
@zhejiangxiaomai 

I think we can simplify the code by using recursion like so:

```
    case ::substrait::Expression_Literal::LiteralTypeCase::kList: {
      VectorPtr elements;
      for (auto it : listLiteral.list().values()) {
        auto v = getInfoFromListLiteral(it);
        if (!elements) {
          elements = v;
        } else {
          elements->append(v.get());
        }
      }

      return makeArrayVector(elements);
    }
```

See the full code in my branch (Review Comments commit):

https://github.com/facebookincubator/velox/compare/main...mbasmanova:velox-1:support_list?expand=1

Reply:
@zhejiangxiaomai I'm finished reviewing this PR. I'll wait for you to consider the comments and reply. If it is going to take you some time to update the PR, please, give me an idea of how long it will take so I can plan better.

Reply:
Thanks, I think I can finish the comments next Tuesday (Nov. 21).

Reply:
Sounds good. I'll be out from next Wed till Dec 1.